### PR TITLE
Pretty print tree 

### PR DESCRIPTION
**Pull request by mathiasvr**
Hi,

I added a pretty print function for #20.
Currently, it's just a quick test to see if this does what you mean?

Sample output:
```
        3
    5
        7
8
        10
    12
        15
```

Do you want it as a traversal strategy, or PrettyPrinter class, or?

### CODE REVIEW

**Comment by Marusyk** [comment]
Hi @mathiasvr,

Thanks for your interest. 
Yes, I mean something like that but 

```
       8
    /    \
  12       5
 /  \    /  \
15  10   7   3
```

In my opinion it would be better to implement BinaryTreePrinter that will print to console according to  traversal strategy

File: src/BinaryTree/BinaryTree.cs [hunk 1]
```diff
@@ -33,6 +33,23 @@ public BinaryTree(int capacity)
         public BinaryTree()
         {
         }
+        
+        public void PrettyPrint()
+	    {
+		    PrettyPrintRec(this._head, 0);
+	    }
+
+	    private void PrettyPrintRec(BinaryTreeNode<T> node, int spaces)
```
Comment (Marusyk) [inline src/BinaryTree/BinaryTree.cs]: Following to Single responsibility principle we should move that functionality to a new class.

File: src/BinaryTree/BinaryTree.cs [hunk 2]
```diff
@@ -33,6 +33,23 @@ public BinaryTree(int capacity)
...
+	    {
+		    if (node != null)
+		    {
+			    PrettyPrintRec(node.Left, spaces + 1);
+			    for (int i = 0; i < spaces; i++)
```
Comment (Marusyk) [inline src/BinaryTree/BinaryTree.cs]: It would be better to use `while`. What do you think about it?

Comment (Marusyk) [reply src/BinaryTree/BinaryTree.cs]: Also please add braces

**Comment by Marusyk** [comment]
Oh, yeah. Sorry for confusing. My bad. There are only one correct way

**Comment by mathiasvr** [comment]
Okay, no problem.
I updated the PR, and added a BinaryTreePrinter class.

I tried to "rotate" the tree, but that was a bit complicated, so I did not complete it.
I hope this simple printer can be used as a starting point.

**Comment by Marusyk** [comment]
@mathiasvr Thanks for contributing, but how can I use your `BinaryTreePrinter`? 

There are only one possibility to use that inside `BinaryTree` class:

```csharp
PrettyPrint(BinaryTreeNode<T> root)PrettyPrint(BinaryTreeNode<T> root)
```

We do not need that because `BinaryTree` should not take any responsibility for printing. 

I expected something like this:

```csharp
var binaryTree = new BinaryTree<int> { 8, 5, 12, 3, 7, 10, 15 };
binaryTree.Print();
```

Also, you need to add some example how to use that in [sample project](https://github.com/Marusyk/BinaryTree/blob/master/samples/BinaryTreeConsole/Program.cs)

**Comment by mathiasvr** [comment]
Not sure I understand it because if BinaryTree has a `Print()` method, it does take responsibility for printing?

Anyway, I updated it and made it pluggable like the Traversal class.

**Comment by Marusyk** [comment]
> Not sure I understand it because if BinaryTree has a Print() method, it does take responsibility for printing?

It should not have! See [this](https://github.com/Marusyk/BinaryTree/issues/20#issue-368188739). It should be done as an extension method

**Comment by mathiasvr** [comment]
Then how do I access the head node?

**Comment by Marusyk** [comment]
Closed because this solution does not meet the requirements

**Pull request closed without merging**